allow positional access in named structs

peterboncz opened the pull request:
allow positional access in named structs

the reason why I propose this is not only to allow more flexible access to structs, but I also would like to fix an issue in TableFilter:ToExpression() which in case of predicates on struct members fails to produce a working boolean expression (because, it creates a positional access rather than a key-based one).

It actually also got that position wrong (off by one).

Commit:
allow positional access in named structs

the reason why I propose this is not only to allow more flexible access to structs, but I also
would like to fix an issue in TableFilter:ToExpression() which in case of predicates on struct
members fails to produce a working boolean expression (because, it creates a positional access
rather than a key-based one).

It actually also got the that position wrong (off by one).

Commit:
format-fix (attempt, as I cannot get DDB format-fix to work)

Mytherin commented:
Thanks for the PR!

Perhaps we can have a different method that does this? Overloading the struct extract means it becomes possible to index structs like arrays (i.e. `struct_col[1]`) which I think is undesirable/potentially confusing.

Commit:
forgot this change:
- struct_filter is now OK with positional struct_extract

peterboncz commented:
Right, something like struct_extract_by_position() ?

Two remarks though:
- I suppose st[key] is syntactic sugar for struct_extract(st, key), so defining a new method would mean that the syntactic sugar won't work for the positional access and there is no amboiguity
- However, st[1] that is positional access is already supported now. But it is only supported on structs that have no names for the keys (isUnnamed()) - so would it not be just a matter of consistency to allow it?

Mytherin commented:
Unnamed structs are not really structs, they are more like tuples in Python. They don't have names so elements must be accessed by position. The existing struct extract syntax can keep working for them as it works in python, e.g.:

```py
>>> a = (42, 84);
>>> a[1]
84
>>> a = {'key1': 42, 'key2': 84}
>>> a[1]
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    a[1]
    ~^^^
KeyError: 1

```

Commit:
- implemented extract_at(st, pos)
- added a few test cases
- changed order of StructFilter constructor to make key name last and optional

peterboncz commented:
went for the name struct_extract_at(st, pos) because it is shorter

Commit:
format-fix

Commit:
fix my test

Commit:
make ToString compatible with the new method naming

Commit:
format-fix again
keep original struct_filter ToString the same

Commit:
fixed a bug in my assert

Commit:
giving up on format-fix

Commit:
Merge remote-tracking branch 'origin/master' into pb/struct-unpack-fix

Commit:
the remove_unused_columns rule now must recongize struct_extract_at

Commit:
format-fix in the dark without tools

Commit:
Merge remote-tracking branch 'origin/master' into pb/struct-unpack-fix

Commit:
add json defintion of struct_extract_at()

Commit:
playing detective as what the generated code should look like
(whithout the ability to generate it, ha ha)
almost there, it seems, one CR

Commit:
revert the change of parameter order in StructFilter constructor
- it breaks serialization, and causes my generated-files woes in CI
not worth it

Commit:
use child_name.empty() iso child_name == ""

Mytherin commented on src/function/scalar/struct/struct_extract.cpp:
```diff
@@ -87,8 +87,8 @@ static unique_ptr<FunctionData> StructExtractBind(ClientContext &context, Scalar
-static unique_ptr<FunctionData> StructExtractBindIndex(ClientContext &context, ScalarFunction &bound_function,
-                                                       vector<unique_ptr<Expression>> &arguments) {
+static unique_ptr<FunctionData> StructExtractBindInternal(ClientContext &context, ScalarFunction &bound_function,
+                                                          vector<unique_ptr<Expression>> &arguments, bool by_pos) {
```
Minor nit, but this bool should probably be `bool by_name`

peterboncz replied:
both functions actually do 'by_pos'. 

I renamed the bool 'struct_extract' now to distinguish between the behavior of 'struct_extract()' and 'struct_extract_at()' => the former only allows positional access on unnamed structs.

Mytherin reviewed (commented):
Thanks! LGTM - one minor nit

Commit:
renamed parameter 'struct'extract' to distinguish between
'struct_extract()' and  'struct_extract_at()'

Commit:
manual format-fix
